bakery/checkers/checkers.go

```diff
@@ -4,192 +4,185 @@ package checkers
-// same condition identifier, all of them will be used.
-//
-// The cause of any error returned by a checker will be preserved.
-//
-// Note that because the returned checker implements Checker
-// as well as bakery.FirstPartyChecker, calls to New can be nested.
-// For example, a checker can be easily added to an existing
-// MultiChecker, by doing:
-//
-//	checker := checkers.New(old, another)
-func New(checkers ...Checker) *MultiChecker {
-	return &MultiChecker{
-		checkers: checkers,
+// RegisterStd registers all the standard checkers in the given checker.
+// The standard namespace is also added to ns if not present.
+func RegisterStd(c *Checker) {
+	c.namespace.Register(StdNamespace, "")
+	for cond, check := range allCheckers {
+		c.Register(cond, StdNamespace, check)
 	}
 }
 
-// MultiChecker implements bakery.FirstPartyChecker
-// and Checker for a collection of checkers.
-type MultiChecker struct {
-	// TODO it may be faster to initialize a map, but we'd
-	// be paying the price of creating and initializing
-	// the map vs a few linear scans through a probably-small
-	// slice. Let's wait for some real-world numbers.
-	checkers []Checker
+// New returns a checker with all the standard caveats checkers registered.
+// If ns is nil, a new one will be created.
+// The standard namespace is also added to ns if not present.
+func New(ns *Namespace) *Checker {
+	c := NewEmpty(ns)
+	RegisterStd(c)
+	return c
 }
 
-var errBadCaveat = errgo.Newf("bad caveat")
+// Checker holds a set of checkers for first party caveats.
+// It implements bakery.CheckFirstParty caveat.
+type Checker struct {
+	namespace *Namespace
+	checkers  map[string]CheckerInfo
+}
 
-// Check implements Checker.Check.
-func (c *MultiChecker) Check(cond, arg string) error {
-	// Always check for the error caveat so that we're
-	// sure to get a nice error message even when there
-	// are no other checkers. This also prevents someone
-	// from inadvertently overriding the error condition.
-	if cond == CondError {
-		return errBadCaveat
+// Register registers the given condition in the given namespace URI
+// to be checked with the given check function.
+// It will panic if the namespace is not registered or
+// if the condition has been registered twice.
+func (c *Checker) Register(cond, uri string, check Func) {
+	if check == nil {
+		panic(fmt.Errorf("nil check function registered for namespace %q when registering condition %q", uri, cond))
 	}
-	checked := false
-	for _, c := range c.checkers {
-		checkerCond := c.Condition()
-		if checkerCond != "" && checkerCond != cond {
-			continue
-		}
-		if err := c.Check(cond, arg); err != nil {
-			if checkerCond == "" && errgo.Cause(err) == ErrCaveatNotRecognized {
-				continue
-			}
-			return errgo.Mask(err, errgo.Any)
-		}
-		checked = true
+	prefix, ok := c.namespace.Resolve(uri)
+	if !ok {
+		panic(fmt.Errorf("no prefix registered for namespace %q when registering condition %q", uri, cond))
```

Comment:
The panics are by design - the registration stuff is designed to happen up front (usually at init time), and if you're registering a duplicate name or haven't registered your URI, I think it's OK to panic, as that's entirely under your control.

If you're making a checker as a global variable (the way I want them to be able to be used), you'll end up panicking anyway if there's you get an error.